src/Features/LanguageServer/Protocol/Handler/InlineCompletions/InlineCompletionsHandler.ParsedXmlSnippet.cs

```diff
@@ -0,0 +1,87 @@
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.InlineCompletions;
+
+internal partial class InlineCompletionsHandler
+{
+    private class ParsedXmlSnippet
+    {
+        public ImmutableArray<SnippetPart> Parts { get; }
+        public string DefaultText { get; }
+
+        public ParsedXmlSnippet(ImmutableArray<SnippetPart> parts)
+        {
+            Parts = parts;
+
+            using var _ = PooledStringBuilder.GetInstance(out var defaultSnippetBuilder);
+            foreach (var part in parts)
+            {
+                var textToAdd = part.DefaultText;
+                defaultSnippetBuilder.Append(textToAdd);
+            }
+
+            DefaultText = defaultSnippetBuilder.ToString();
+        }
+    }
+
+    private abstract record SnippetPart(string DefaultText);
+
+    private record SnippetFieldPart(string FieldName, string DefaultText, int? EditIndex) : SnippetPart(DefaultText);
+
+    private record SnippetFunctionPart(string FieldName, string DefaultText, int? EditIndex, string FunctionName, string? FunctionParam)
+        : SnippetFieldPart(FieldName, DefaultText, EditIndex)
+    {
+        public async Task<SnippetFunctionPart> WithSnippetFunctionResultAsync(Document documentWithSnippet, TextSpan fieldSpan, CancellationToken cancellationToken)
+        {
+            var snippetFunctionService = documentWithSnippet.Project.GetRequiredLanguageService<SnippetFunctionService>();
+            switch (FunctionName)
+            {
+                case "SimpleTypeName":
+                    if (FunctionParam == null)
+                    {
+                        return this;
+                    }
+
+                    var simplifiedTypeName = await SnippetFunctionService.GetSimplifiedTypeNameAsync(documentWithSnippet, fieldSpan, FunctionParam, cancellationToken).ConfigureAwait(false);
+                    if (simplifiedTypeName == null)
+                    {
+                        return this;
+                    }
+
+                    return this with { DefaultText = simplifiedTypeName };
+                case "ClassName":
+                    var className = await snippetFunctionService.GetContainingClassNameAsync(documentWithSnippet, fieldSpan.Start, cancellationToken).ConfigureAwait(false);
+                    if (className == null)
+                    {
+                        return this;
+                    }
+
+                    return this with { DefaultText = className };
+                case "GenerateSwitchCases":
+                    // Generate switch cases requires a multi-step snippet interaction, where the snippet is inserted first then the
+                    // client calls back to on commit so that we can generate the cases for the specified switch value.
+                    // This is not yet supported via LSP.
+                    return this;
```

Comment:
Didn't realize this is a two step process at the time of writing the LSP api :(
Future work here is to figure out a way for LSP to call back on the 'commit' of a tab stop and ask the server for more edits.

Reply:
Interesting, is this also handled in the new semantic snippets?

Reply:
semantic snippets would have to support it in some manner - the cases can only be known when the field is committed